src/stdlib_math_diff.fypp

```diff
@@ -0,0 +1,102 @@
+#! Originally written in 2016 by Keurfon Luu (keurfonluu@outlook.com)
+#! https://github.com/keurfonluu/Forlab/blob/master/src/lib/forlab.f90#L2673
+
+#:include "common.fypp"
+#:set RI_KINDS_TYPES = REAL_KINDS_TYPES + INT_KINDS_TYPES
+submodule (stdlib_math) stdlib_math_diff
+
+    implicit none
+
+contains
+
+    #! `diff` computes differences of arrays of the ${t1}$ type.
+    
+    #:for k1, t1 in RI_KINDS_TYPES
+    pure module function diff_1_${k1}$(X, n, prepend, append) result(Y)
+        ${t1}$, intent(in) :: X(:)
+        integer, intent(in), optional :: n
+        ${t1}$, intent(in), optional :: prepend(:), append(:)
+        ${t1}$, allocatable :: Y(:)
+        integer :: size_prepend, size_append
+        integer :: n_, i
+
+        size_prepend = 0
+        size_append = 0
+        n_ = optval(n, 1)
+        if (present(prepend)) size_prepend = size(prepend) 
+        if (present(append)) size_append = size(append)
+        
+        if (size(X) + size_prepend + size_append <= n_) then
+            allocate(Y(0))
+            return
+        end if
+
+        if (size_prepend > 0) then
+            Y = [prepend, X]
+        else
+            Y = X
+        end if
+        if (size_append > 0) Y = [Y, append]
+        
+        do i = 1, n_
+            Y = Y(2:) - Y(:size(Y) - 1)
+        end do
```

Comment:
I think there's another difference with numpy here. In the latter, the output size is `(size(x) + size(prepend) + size(append) - n)`, for example:
```python
In [1]: import numpy

In [2]: numpy.diff([1, 4, 7, 20])
Out[2]: array([ 3,  3, 13])

In [3]: numpy.diff([1, 4, 7, 20], n=2)
Out[3]: array([ 0, 10])

# Cases using prepend and append

In [4]: numpy.diff([1, 4, 7, 20], n=2, prepend=[1,1])
Out[4]: array([ 0,  3,  0, 10])

In [5]: numpy.diff([1, 4, 7, 20], n=2, prepend=[1,1], append=[50, 50])
Out[5]: array([  0,   3,   0,  10,  17, -30])

```

Reply:
The feature of the allocatable array is used here, and the length of the array is updated according to the result each time the loop is looped.
```fortran
integer :: x(8) = [1, 1, 1, 4, 7, 20, 50, 50]
integer, allocatable :: y(:)

y = x
do i = 1, 2
    Y = Y(2:) - Y(:size(Y) - 1)
end do

print "(1x, *(i0, 2x))", Y  !! [ 0,  3,  0,  10,  17,  -30 ]

end
```

Reply:
Oh my mistake, interesting!

Reply:
It would be good to understand whether all the reallocations lead to performance hits when `n` is large. An alternative approach would be to use `Y(2:) = Y(2:) - Y(:size(Y)-1)` in the loop, and then after the loop, have a line `Y = Y(n_+1:)`